### PR TITLE
Dilation2D Op Support

### PR DESCRIPTION
**Pull request by maheshambule**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Comment by hcho3** [comment]
CI is back online. Re-triggering tests.

File: python/tvm/relay/frontend/tensorflow.py [hunk 1]
```diff
@@ -425,6 +427,106 @@ def _impl(inputs, attr, params):
         return out
     return _impl
 
+
+#Dilation2d
+def _dilation2d(opname):
+
+    def _impl(inputs, attr, params):
+        attr['data_format'] = 'NHWC'
```
Comment (yongwww) [inline python/tvm/relay/frontend/tensorflow.py:435]: why does the layout hard-coded here?

File: python/tvm/relay/frontend/tensorflow.py [hunk 2]
```diff
@@ -425,6 +427,106 @@ def _impl(inputs, attr, params):
...
+                inputs[0] = _op.nn.pad(data=inputs[0],
+                                       pad_width=((0, 0),
+                                                  (0, 0),
+                                                  (pad_v[0], pad_v[1]),
+                                                  (pad_h[0], pad_h[1])))
+
+            attr['padding'] = [0, 0]
+
+        else:
+            msg = 'Value {} in attribute "padding" of operator Conv is not ' \
```
Comment (yongwww) [inline python/tvm/relay/frontend/tensorflow.py]: topo Conv

File: python/tvm/relay/frontend/tensorflow.py [hunk 3]
```diff
@@ -425,6 +427,106 @@ def _impl(inputs, attr, params):
...
+
+        attr['data_format'] = 'NCHW'
+
+        # Ignore the new attributes from TF2.0, for now.
+        out = AttrCvt(
+            op_name='dilation2d',
+            ignores=['explicit_paddings'],
+            transforms={
+                'data_format': 'data_layout',
+
```
Comment (yongwww) [inline python/tvm/relay/frontend/tensorflow.py:522]: remove the empty line

File: python/tvm/relay/op/nn/_nn.py [hunk 4]
```diff
@@ -389,6 +389,40 @@ def legalize_conv2d_transpose(attrs, inputs, types):
...
+    out_dtype = attrs.out_dtype
+    out_dtype = (inputs[0].dtype if out_dtype in ("same", "")
+                 else out_dtype)
+
+    assert layout in ["NCHW", "NHWC", "HWCN"]
+    (dilation_h, dilation_w) = rates
+    if dilation_h < 1 or dilation_w < 1:
+        raise ValueError("dilation should be positive value")
+
+
```
Comment (yongwww) [inline python/tvm/relay/op/nn/_nn.py]: don't need the empty line

File: python/tvm/relay/op/nn/_nn.py [hunk 5]
```diff
@@ -389,6 +389,40 @@ def legalize_conv2d_transpose(attrs, inputs, types):
...
+    out = topi.nn.dilation2d(
+        inputs[0], inputs[1], strides, padding,
+        rates, layout, out_dtype)
+
```
Comment (yongwww) [inline python/tvm/relay/op/nn/_nn.py]: remove this line

File: python/tvm/relay/op/nn/_nn.py [hunk 6]
```diff
@@ -389,6 +389,40 @@ def legalize_conv2d_transpose(attrs, inputs, types):
...
+    return [out]
+
+
+@reg.register_schedule("nn.dilation2d")
+def schedule_dilation2d(attrs, outs, target):
+    """Schedule definition of dilation2d"""
+    return topi.generic.schedule_dilation2d(outs)
+
```
Comment (yongwww) [inline python/tvm/relay/op/nn/_nn.py:418]: remove this line

File: src/relay/op/nn/convolution.cc [hunk 7]
```diff
@@ -1099,5 +1099,77 @@ TVM_REGISTER_API("relay.op.nn._make.deformable_conv2d")
...
+  // So this inference ignores all inputs
+  return Array<Array<Layout> >{{params->data_layout, params->kernel_layout},
+                               {params->out_layout == "" ?
+                                   params->data_layout : params->out_layout}};
+}
+
+// Positional relay function to create dilation2d operator
+// used by frontend FFI.
+Expr MakeDilation2D(Expr data,
+                Expr weight,
```
Comment (yongwww) [inline src/relay/op/nn/convolution.cc]: indentation

File: tests/python/frontend/tensorflow/test_forward.py [hunk 8]
```diff
@@ -2866,10 +2866,49 @@ def test_forward_add_n():
...
+                              in_filter,
+                              strides=strides,
+                              rates=rates,
+                              padding=padding)
+
+            compare_tf_with_tvm(np.reshape(data_array, tensor_in_sizes).astype('float32'),
+                                'Placeholder:0', 'Dilation2D:0')
+
+def test_forward_dilation():
+    if is_gpu_available():
```
Comment (yongwww) [inline tests/python/frontend/tensorflow/test_forward.py]: is this op only for GPU?

File: tests/python/frontend/tensorflow/test_forward.py [hunk 9]
```diff
@@ -2866,10 +2866,49 @@ def test_forward_add_n():
     _test_forward_add_n(in5)
 
 
+def _test_dilation2d(opname, tensor_in_sizes, filter_in_sizes,
```
Comment (yongwww) [inline tests/python/frontend/tensorflow/test_forward.py]: no need to use an extra arg opname

File: tests/python/relay/test_op_level2.py [hunk 10]
```diff
@@ -987,6 +1052,8 @@ def test_bitpack_infer_type():
 
 
 if __name__ == "__main__":
+    test_dilation2d_infer_type()
```
Comment (yongwww) [inline tests/python/relay/test_op_level2.py]: merge this two together and use test_dilation2d instead

File: topi/python/topi/testing/dilation2d_python.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (yongwww) [inline topi/python/topi/testing/dilation2d_python.py]: merge the func in dilate_python.py

File: topi/python/topi/nn/dilation2d.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (yongwww) [inline topi/python/topi/nn/dilation2d.py]: merge it in dilate.py?

**Review by yongwww — CHANGES_REQUESTED** [review]

File: tests/python/relay/test_op_level2.py (hunk 10 above)
Comment (mjpsl) [reply tests/python/relay/test_op_level2.py]: @yongwww Reason to have two separate tests is, we are just following conv2d operator convention where there are separate tests for infer type and operator output testing. do you still want to merge test case

Comment (yongwww) [reply tests/python/relay/test_op_level2.py]: let's merge them, at least it looks simpler.

File: python/tvm/relay/op/nn/_nn.py (hunk 4 above)
Comment (mjpsl) [reply python/tvm/relay/op/nn/_nn.py]: removed empty lines

File: python/tvm/relay/frontend/tensorflow.py (hunk 3 above)
Comment (mjpsl) [reply python/tvm/relay/frontend/tensorflow.py:522]: removed empty lines

File: python/tvm/relay/op/nn/_nn.py (hunk 6 above)
Comment (mjpsl) [reply python/tvm/relay/op/nn/_nn.py:418]: removed empty lines

File: python/tvm/relay/frontend/tensorflow.py (hunk 1 above)
Comment (mjpsl) [reply python/tvm/relay/frontend/tensorflow.py:435]: As per the TF 1.14 , there is no separate attribute data_format for operator dilation2d, so it was not passed /processed in attribute list. but for further processing we needed this, that is why we have introduced it.

File: python/tvm/relay/frontend/tensorflow.py (hunk 2 above)
Comment (mjpsl) [reply python/tvm/relay/frontend/tensorflow.py]: corrected the typo error

File: python/tvm/relay/op/nn/_nn.py (hunk 5 above)
Comment (mjpsl) [reply python/tvm/relay/op/nn/_nn.py]: removed empty line

File: src/relay/op/nn/convolution.cc (hunk 7 above)
Comment (mjpsl) [reply src/relay/op/nn/convolution.cc]: corrected the indentation

File: tests/python/frontend/tensorflow/test_forward.py (hunk 8 above)
Comment (mjpsl) [reply tests/python/frontend/tensorflow/test_forward.py]: this is no only for GPU, so changed the execution plan without GPU. CI is through now

File: tests/python/frontend/tensorflow/test_forward.py (hunk 9 above)
Comment (mjpsl) [reply tests/python/frontend/tensorflow/test_forward.py]: removed the extra arg opname

File: tests/python/relay/test_op_level2.py (hunk 10 above)
Comment (mjpsl) [reply tests/python/relay/test_op_level2.py]: merged

File: topi/python/topi/testing/dilation2d_python.py (hunk 11 above)
Comment (mjpsl) [reply topi/python/topi/testing/dilation2d_python.py]: merged in dilate_python.py

File: topi/python/topi/nn/dilation2d.py (hunk 12 above)
Comment (mjpsl) [reply topi/python/topi/nn/dilation2d.py]: merged in dilate.py, file deleted dilation2d.py

File: python/tvm/relay/frontend/tensorflow.py (hunk 1 above)
Comment (mjpsl) [reply python/tvm/relay/frontend/tensorflow.py:435]: > why does the layout hard-coded here?
Hello Yong, not sure if you got to see my response. could you please confirm , if comment is fine?

**Comment by maheshambule** [comment]
PR created against Apache. Hence closing.

**Pull request closed without merging**